cpp/src/arrow/compute/kernels/scalar_hash_test.cc

```diff
@@ -0,0 +1,336 @@
+    CheckDeterminisic(func, arr);
+    if (func == "hash64") {
+      AssertHashesEqual<uint64_t>(arr, hash_result, HashBinaryLike<uint64_t>(arr));
+    } else if (func == "hash32") {
+      AssertHashesEqual<uint32_t>(arr, hash_result, HashBinaryLike<uint32_t>(arr));
+    } else {
+      FAIL() << "Unknown function: " << func;
+    }
+  }
+};
+
+TEST_F(TestScalarHash, Null) {
+  Datum res;
+  std::shared_ptr<Array> arr;
+  std::shared_ptr<Array> exp;
+
+  arr = ArrayFromJSON(null(), R"([])");
+  exp = ArrayFromJSON(uint32(), "[]");
+  ASSERT_OK_AND_ASSIGN(res, CallFunction("hash32", {arr}));
+  AssertArraysEqual(*res.make_array(), *exp);
+  CheckDeterminisic("hash32", arr);
+
+  arr = ArrayFromJSON(null(), R"([])");
+  exp = ArrayFromJSON(uint64(), "[]");
+  ASSERT_OK_AND_ASSIGN(res, CallFunction("hash64", {arr}));
+  AssertArraysEqual(*res.make_array(), *exp);
+  CheckDeterminisic("hash64", arr);
+
+  arr = ArrayFromJSON(null(), R"([null, null, null])");
+  exp = ArrayFromJSON(uint32(), "[0, 0, 0]");
+  ASSERT_OK_AND_ASSIGN(res, CallFunction("hash32", {arr}));
+  AssertArraysEqual(*res.make_array(), *exp);
+  CheckDeterminisic("hash32", arr);
+
+  arr = ArrayFromJSON(null(), R"([null, null, null])");
+  exp = ArrayFromJSON(uint64(), "[0, 0, 0]");
+  ASSERT_OK_AND_ASSIGN(res, CallFunction("hash64", {arr}));
+  AssertArraysEqual(*res.make_array(), *exp);
+  CheckDeterminisic("hash64", arr);
+}
+
+TEST_F(TestScalarHash, NullHashIsZero) {
+  auto arr1 = ArrayFromJSON(int32(), R"([null, 0, 1])");
+  ASSERT_OK_AND_ASSIGN(auto res1, CallFunction("hash64", {arr1}));
+  auto buf1 = res1.array()->GetValues<uint64_t>(1);
+  ASSERT_EQ(buf1[0], 0);
+  ASSERT_NE(buf1[1], 0);
+  ASSERT_NE(buf1[2], 0);
+  ASSERT_NE(buf1[1], buf1[2]);
+
+  auto arr2 = ArrayFromJSON(int8(), R"([null, 0, 1])");
+  ASSERT_OK_AND_ASSIGN(auto res2, CallFunction("hash32", {arr2}));
+  auto buf2 = res2.array()->GetValues<uint32_t>(1);
+  ASSERT_EQ(buf2[0], 0);
+  ASSERT_NE(buf2[1], 0);
+  ASSERT_NE(buf2[2], 0);
+  ASSERT_NE(buf2[1], buf2[2]);
+}
+
+TEST_F(TestScalarHash, Boolean) {
+  auto arr = ArrayFromJSON(boolean(), R"([true, false, null, true, null])");
+  CheckDeterminisic("hash32", arr);
+  CheckDeterminisic("hash64", arr);
+}
+
+TEST_F(TestScalarHash, NumericLike) {
+  auto types = {int8(),   int16(),  int32(),  int64(),   uint8(),
+                uint16(), uint32(), uint64(), float32(), float64()};
+  for (auto func : {"hash32", "hash64"}) {
+    for (auto type : types) {
+      CheckPrimitive(func, ArrayFromJSON(type, R"([])"));
+      CheckPrimitive(func, ArrayFromJSON(type, R"([null])"));
+      CheckPrimitive(func, ArrayFromJSON(type, R"([1])"));
+      CheckPrimitive(func, ArrayFromJSON(type, R"([1, 2, null])"));
+      CheckPrimitive(func, ArrayFromJSON(type, R"([null, 2, 3])"));
+      CheckPrimitive(func, ArrayFromJSON(type, R"([1, 2, 3, 4])"));
+    }
+  }
+}
+
+TEST_F(TestScalarHash, BinaryLike) {
+  auto types = {binary(), utf8(), large_binary(), large_utf8()};
```

Comment:
Does it work for string-view types as well? And fixed-size binary?

Reply:
Added tests for `fixed_size_binary()`. Views are unsupported since the underlying hashing utility only support a fixedLen and a varLen buffer, we cannot pass all the buffers of a string view. Perhaps we could convert it to an ordinary string array first, but that sounds expensive.

Reply:
Yes, we should use another approach if we want to support it (that can be done later).